File: go/metrics/metrics_test.go
package metrics

import (
	"context"
	"testing"

	"github.com/prometheus/client_golang/prometheus/testutil"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/opensourceways/obs-sdk/go/sdkctx"
)

func noProcess() *bool { b := false; return &b }

func newTest(t *testing.T, cfg Config) *Metrics {
	t.Helper()
	cfg.IncludeProcess = noProcess()
	return New(cfg)
}

func TestConstLabelsPresent(t *testing.T) {
	m := newTest(t, Config{Service: "review", Env: "test", Instance: "pod-1", Community: "openeuler"})
	c := m.NewCounterVec("events_total", "events")
	c.Inc()

	fam, err := m.Registry().Gather()
	require.NoError(t, err)
	require.Len(t, fam, 1)

	series := fam[0].GetMetric()
	require.Len(t, series, 1)

	seen := map[string]string{}
	for _, lp := range series[0].GetLabel() {
		seen[lp.GetName()] = lp.GetValue()
	}
	assert.Equal(t, "review", seen["service"])
	assert.Equal(t, "test", seen["env"])
	assert.Equal(t, "pod-1", seen["instance"])
	assert.Equal(t, "openeuler", seen["community"])
	assert.Equal(t, "events_total", fam[0].GetName())
}

func TestDefaultCommunityValue(t *testing.T) {
	m := newTest(t, Config{Service: "review", Community: "openeuler"})
	c := m.NewCounterVec("events_total", "events")
	c.Inc()

	// 打点用部署级默认 community。
	assert.Equal(t, float64(1), testutil.ToFloat64(
		c.Vec().WithLabelValues("openeuler")))
}

func TestDynamicCommunityOverride(t *testing.T) {
	m := newTest(t, Config{Service: "review", Community: "openeuler"})
	c := m.NewCounterVec("events_total", "events", "kind")

	c.Add(2, "pr")                          // 默认 openeuler
	c.IncWithContext(ctx("mindspore"), "pr") // 覆盖 mindspore

	assert.Equal(t, float64(2), testutil.ToFloat64(c.Vec().WithLabelValues("openeuler", "pr")))
	assert.Equal(t, float64(1), testutil.ToFloat64(c.Vec().WithLabelValues("mindspore", "pr")))

	// 覆盖只影响打点时刻，不污染后续默认值。
	c.Add(1, "issue")
	assert.Equal(t, float64(1), testutil.ToFloat64(c.Vec().WithLabelValues("openeuler", "issue")))
}

func TestGauge(t *testing.T) {
	m := newTest(t, Config{Service: "review", Community: "openeuler"})
	g := m.NewGaugeVec("in_flight", "in flight")
	g.Set(3)
	g.Inc()
	assert.Equal(t, float64(4), testutil.ToFloat64(g.Vec().WithLabelValues("openeuler")))
}

func TestHistogramCount(t *testing.T) {
	m := newTest(t, Config{Service: "review", Community: "openeuler"})
	h := m.NewHistogramVec("request_duration_seconds", "latency", "method")
	h.Observe(0.1, "GET")
	h.Observe(0.2, "GET")

	// histogram 不能 testutil.ToFloat64 直接读，改按 family 校验 _count。
	fam, err := m.Registry().Gather()
	require.NoError(t, err)
	require.Len(t, fam, 1)
	mf := fam[0].GetMetric()
	require.Len(t, mf, 1)
	assert.Equal(t, uint64(2), mf[0].GetHistogram().GetSampleCount())
	assert.Equal(t, "request_duration_seconds", fam[0].GetName())
}

func TestFQNameNamespace(t *testing.T) {
	m := newTest(t, Config{Service: "review", Namespace: "obs"})
	c := m.NewCounterVec("events_total", "events")
	c.Inc()

	fam, err := m.Registry().Gather()
	require.NoError(t, err)
	require.Equal(t, "obs_events_total", fam[0].GetName())
}

func ctx(community string) context.Context {
	return sdkctx.WithCommunity(context.Background(), community)
}

File: node/test/metrics.test.js
'use strict';

const { test } = require('node:test');
const assert = require('node:assert');
const { Metrics } = require('../lib/metrics');
const context = require('../lib/context');

test('counter 带公共 label + community 双层注入', async () => {
  const m = new Metrics({ service: 'review', env: 'test', instance: 'pod-1',
    community: 'openeuler' });
  const c = m.counter('events_total', 'events', ['kind']);
  c.inc(1, { kind: 'pr' });
  context.bindRequest({ community: 'mindspore' }, () => {
    c.inc(1, { kind: 'pr' });
  });

  const text = await m.text();
  const rows = text.split('\n').filter((l) => /^events_total\{/.test(l));
  assert.strictEqual(rows.length, 2);
  const byComm = {};
  for (const r of rows) {
    const matched = r.match(/community="([^"]+)"/);
    byComm[matched[1]] = r;
  }
  assert.ok(byComm.openeuler.endsWith(' 1'));
  assert.ok(byComm.mindspore.endsWith(' 1'));
  assert.ok(byComm.openeuler.includes('service="review"'));
});

test('gauge + histogram', async () => {
  const m = new Metrics({ service: 'review', community: 'openeuler' });
  const g = m.gauge('in_flight', 'in flight');
  g.set(3);
  g.inc(2);
  const h = m.histogram('latency_seconds', 'latency');
  h.observe(0.1);
  h.observe(0.2);

  const text = await m.text();
  const gaugeRows = text.split('\n').filter((l) => /^in_flight\{/.test(l));
  assert.ok(gaugeRows[0].endsWith(' 5'));
  assert.ok(text.includes('latency_seconds_count'));
  const countRows = text.split('\n').filter((l) => /^latency_seconds_count\{/.test(l));
  assert.ok(countRows[0].endsWith(' 2'));
});

test('namespace 前缀', async () => {
  const m = new Metrics({ service: 'review', namespace: 'obs' });
  m.counter('events_total', 'events').inc(1);
  const text = await m.text();
  assert.ok(text.includes('obs_events_total'));
});
